### ISSUE TITLE
Allow schema manipulation via 3rd-party

### ISSUE THREAD

**Issue by timkelty** [markhuot/craftql#134]
I want to be able to add arbitrary things to the GraphQL schema.

I _think_ adding an event in `markhuot\CraftQL\Services\getSchema` is all I'm looking for, but I'm likely over-simplifying things: https://github.com/markhuot/craftql/blob/master/src/Services/GraphQLService.php#L118

That way a 3rd party dev and can whatever they might want.

Would be a way for a plugin to add non-element/field based data to the schema, e.g. https://github.com/markhuot/craftql/issues/102, https://github.com/markhuot/craftql/issues/124

**Comment by narration-sd** [comment]
This seems a very sensible request -- in particular because it should reduce Mark Huot's workload or responsibility as far as feature adds.

It would also provide a good proving path for ideas which might progress to FRs or PRs.

**Comment by khalwat** [comment]
This is something I want, too. I'd  like to be able to add support for querying SEOmatic meta containers via GraphQL/CraftQL. The thing is, although there _is_ and SEO Settings field, this isn't what returns the data people will want.

Currently it can be done by pinging a controller as per:

https://github.com/nystudio107/craft-seomatic#headless-spa-api

...where you pass in a URI, and it returns to you the appropriate meta containers for that URI.

It's unclear to me how I might be able to add support for this functionality from my plugin into CraftQL?

**Comment by markhuot** [comment]
This is now possible on `dev-master`. The way to do it is a new event on the `Query` type. E.g.,

```php
Event::on(
    \markhuot\CraftQL\Types\Query::class,
    \markhuot\CraftQL\Events\AlterQuerySchema::EVENT,
    function (\markhuot\CraftQL\Events\AlterQuerySchema $event) {
        // do whatever modifications to the root schema here.
        $event->query->addStringField('foo')->resolve('bar');
        $event->query->addBooleanField('baz')->resolve(function () { return false; });
    }
);
```

☝️ this would probably go in your Plugin `init` method, if you're going to use it.

This answers a small part of the "how do I customize CraftQL" but it seems like the most flexible place to start. Let me know if this works for all your purposes.

If anyone has a chance to test this out on `dev-master` and provide feedback, we can probably roll it in before it makes it to a stable release.

**Comment by timkelty** [comment]
@markhuot this worked pretty well for me! I roughed in support for the verbb/navigation plugin (#124): https://github.com/verbb/navigation/compare/master...timkelty:wip-craftql?expand=1

This provides an input schema like so:

```
{
  navigation(handle: "footer" site: "website") {
    ...nodeFields
    parent {
      ...nodeFields
    }
    siblings {
      ...nodeFields
    }
    children {
      ...nodeFields
    }
    ancestors {
      ...nodeFields
    }
  }
}

fragment nodeFields on NavigationNodeInterface {
  id
  title
  url
  enabled
  status
  level
  elementId
  newWindow
}
```

I mostly just fumbled through it, looking at what you did and trying to replicate. If you care to have a look, let me know if anything looks off.

#### I still have a few questions and fuzzy areas…

- `$context, $info` in resolve callback…should I be using these? If so, how?
- I'm a bit fuzzy on what my `markhuot\CraftQL\Builders\Schema::getName` should be. should it be dynamic somehow based on the context?
- Same as above for `markhuot\CraftQL\Builders\InterfaceBuilder::getResolveType`

#### Other things I'd like it to do but couldn't figure out

- I want the `element` field of `NavigationNodeInterface` to be whatever type the element is, so instead of just `ElementInterface`, it would be `EntryInterface` or `VolumeInteface`, etc.
- Ultimately, i'd like to structure the `navigation` field to return top level info, (`id`, `handle`, etc) and then have a `nodes` subfield that returns the nodes (optionally accepting args). I wasn't sure how to have that `nodes` field only available within `navigation`, and how to pass down arg vals from `navigation`.

**Comment by khalwat** [comment]
This is high on my list as per https://github.com/nystudio107/craft-seomatic/issues/247

**Comment by markhuot** [comment]
`\markhuot\CraftQL\Events\AlterQuerySchema::EVENT` is now included in version 1.2.1, I'm going to close this issue although there is still a good amount of work going in to the 2.x branch that will improve this even more.
